src/node_url.cc

```diff
@@ -1351,6 +1351,53 @@ namespace url {
                             v8::NewStringType::kNormal).ToLocalChecked());
   }
 
+  static void ToUSVString(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+    CHECK_GE(args.Length(), 1);
+    Local<String> str;
+    if (!args[0]->ToString(env->context()).ToLocal(&str))
+      return;
+
+    // The algorithm below can be replaced with `Utf8Value`, which
+    // automatically replaces invalid Unicode surrogates with U+FFFD
+    // REPLACEMENT CHARACTER as the Web IDL spec mandates through V8's
+    // `String::REPLACE_INVALID_UTF8`. However, using `Utf8Value` would prevent
+    // us from using the `changed` optimization, and V8 would have to create a
+    // new string even when it is not needed.
+
+    TwoByteValue value(env->isolate(), str);
+    const size_t n = value.length();
+    bool changed = false;
+
+    for (size_t i = 0; i < n; i++) {
+      uint16_t c = value[i];
+      if (c < 0xD800 || c > 0xDFFF) {
+        continue;
+      } else if (0xDC00 <= c || i == n - 1) {
+        changed = true;
+        value[i] = 0xFFFD;
+      } else {
+        uint16_t d = value[i + 1];
+        if (0xDC00 <= d && d <= 0xDFFF) {
+          i++;
+        } else {
+          changed = true;
+          value[i] = 0xFFFD;
+        }
+      }
+    }
```

Comment:
@joyeecheung, I don't think your code snippet would work for a "high high low" sequence, since in the process of checking for "high" `i` is advanced twice, so it would go directly to "low" for the next iteration.

@bnoordhuis, indeed a regex is a lot faster than iterating through the string in JS. The seemingly long regex doesn't actually have much adverse performance effect if the last `([^\uDC00-\uDFFF]|$)` is turned into a negative lookahead. PTAL.